File: src/matlab/com/springbok/pattern/PatternENOR208V01Test.m
classdef PatternENOR208V01Test < TestUtility
% Tests methods of PatternENOR208V01 class.
  
% Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved.

  properties (Constant = true)
    
    GainMax_input_1 = 57;
    Phi0_input = 2;
    
    self_1 = struct( ...
        'GainMax', 57, ...
        'Phi0', 2);
    
    Phi_input = [0.2; 1; 10; 150];
    
    G_expected_1 = [ ...
        57.000000000000000
        54.000000000000000
        27.025749891599531
        15.000000000000000
                   ];
    Gx_expected_1 = [ ...
        57.000000000000000
        54.000000000000000
        27.025749891599531
        15.000000000000000
                    ];
    
    GainMax_input_2 = 47;
    
    self_2 = struct( ...
        'GainMax', 47, ...
        'Phi0', 2);
    
    G_expected_2 = [ ...
        47.000000000000000
        44.000000000000000
        17.025749891599531
        5.000000000000000
                   ];
    Gx_expected_2 = [ ...
        47.000000000000000
        44.000000000000000
        17.025749891599531
        5.000000000000000
                    ];
    
  end % properties (Constant = true)
  
  methods
    
    function this = PatternENOR208V01Test(logFId, testMode)
    % Constructs a PatternENOR208V01Test
    %
    % Parameters:
    %   logFId - Log file identifier
    %   testMode - Test mode, if 'interactive' then beeps and pauses
      
    % Invoke superclass constructor
      if nargin == 0
        superArgs = {};
        
      else
        superArgs{1} = logFId;
        superArgs{2} = testMode;
        
      end % if
      this@TestUtility(superArgs{:});
      
    end % PatternENOR208V01Test()
    
    function test_PatternENOR208V01(this)
    % Tests PatternENOR208V01 method.
      
      p_1 = PatternENOR208V01(this.GainMax_input_1, this.Phi0_input);
      
      this.assert_true( ...
          'PatternENOR208V01', ...
          sprintf('PatternENOR208V01:%.1f', this.GainMax_input_1), ...
          this.HIGH_PRECISION_DESC, ...
          PatternUtility.compare_pattern(this.self_1, p_1));
      
      p_2 = PatternENOR208V01(this.GainMax_input_2, this.Phi0_input);
      
      this.assert_true( ...
          'PatternENOR208V01', ...
          sprintf('PatternENOR208V01:%.1f', this.GainMax_input_2), ...
          this.HIGH_PRECISION_DESC, ...
          PatternUtility.compare_pattern(this.self_2, p_2));
      
    end % test_PatternENOR208V01()
    
    function test_gain(this)
    % Tests gain method.
      
      warning('off');
      
      p_1 = PatternENOR208V01(this.GainMax_input_1, this.Phi0_input);
      
      [G_actual, Gx_actual] = p_1.gain(this.Phi_input);
      
      t(1) = max(abs(G_actual - this.G_expected_1) < this.HIGH_PRECISION);
      t(2) = max(abs(Gx_actual - this.Gx_expected_1) < this.HIGH_PRECISION);
      
      this.assert_true( ...
          'PatternENOR208V01', ...
          sprintf('gain:%.1f', this.GainMax_input_1), ...
          this.HIGH_PRECISION_DESC, ...
          min(t));
      
      [G, Gx, fH] = p_1.gain([0:.001:180], 'PlotFlag', true);
      pause(1); close(fH);
      
      p_2 = PatternENOR208V01(this.GainMax_input_2, this.Phi0_input);
      
      [G_actual, Gx_actual] = p_2.gain(this.Phi_input);
      
      t(1) = max(abs(G_actual - this.G_expected_2) < this.HIGH_PRECISION);
      t(2) = max(abs(Gx_actual - this.Gx_expected_2) < this.HIGH_PRECISION);
      
      this.assert_true( ...
          'PatternENOR208V01', ...
          sprintf('gain:%.1f', this.GainMax_input_2), ...
          this.HIGH_PRECISION_DESC, ...
          min(t));
      
      [G, Gx, fH] = p_2.gain([0:.001:180], 'PlotFlag', true);
      pause(1); close(fH);
      
      warning('on');
      
    end % test_gain()
    
  end % methods
  
end % classdef

File: src/java/com/springbok/pattern/PatternENOR208V01Test.java
/* Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved. */

package com.springbok.pattern;

import com.springbok.utility.TestUtility;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertTrue;

/**
 * Tests methods of PatternENOR208V01 class.
 */
public class PatternENOR208V01Test {
	private static final double GainMax_input_1 = 57;
	private static final double Phi0_input = 2;
	private static final double[] Phi_input = new double[] { 0.2, 1.0, 10.0, 150.0 };
	private static final double[] G_expected_1 = new double[] { 57.0, 54.0, 27.025749891599531, 15.0 };
	private static final double[] Gx_expected_1 = new double[] { 57.0, 54.0, 27.025749891599531, 15.0 };
	private static final double GainMax_input_2 = 47;
	private static final double[] G_expected_2 = new double[] { 47.0, 44.0, 17.025749891599531, 5.0 };
	private static final double[] Gx_expected_2 = new double[] { 47.0, 44.0, 17.025749891599531, 5.0 };

	/**
	 * Tests PatternENOR208V01 method.
	 */
	@Test
	public void test_PatternENOR208V01() {
		PatternENOR208V01 p_1 = new PatternENOR208V01(GainMax_input_1, Phi0_input);

		assertTrue(TestUtility.isDoublesEquals(57, p_1.getGainMax()));
		assertTrue(TestUtility.isDoublesEquals(2, p_1.getPhi0()));

		PatternENOR208V01 p_2 = new PatternENOR208V01(GainMax_input_2, Phi0_input);

		assertTrue(TestUtility.isDoublesEquals(47, p_2.getGainMax()));
		assertTrue(TestUtility.isDoublesEquals(2, p_2.getPhi0()));
	}

	/**
	 * Tests gain method.
	 */
	@Test
	public void test_gain() {
		PatternENOR208V01 p_1 = new PatternENOR208V01(GainMax_input_1, Phi0_input);

		Gain[] gains = new Gain[Phi_input.length];
		Map<String, Object> options = new HashMap<String, Object>();
		options.put("DoValidate", false);

		for (int i = 0; i < gains.length; i++) {
			Gain g = p_1.gain(Phi_input[i], options);
			gains[i] = g;
		}
		for (int i = 0; i < gains.length; i++) {
			assertTrue(TestUtility.isDoublesEquals(G_expected_1[i], gains[i].G));
			assertTrue(TestUtility.isDoublesEquals(Gx_expected_1[i], gains[i].Gx));
		}

		PatternENOR208V01 p_2 = new PatternENOR208V01(GainMax_input_2, Phi0_input);

		gains = new Gain[Phi_input.length];
		options.put("DoValidate", false);

		for (int i = 0; i < gains.length; i++) {
			Gain g = p_2.gain(Phi_input[i], options);
			gains[i] = g;
		}
		for (int i = 0; i < gains.length; i++) {
			assertTrue(TestUtility.isDoublesEquals(G_expected_2[i], gains[i].G));
			assertTrue(TestUtility.isDoublesEquals(Gx_expected_2[i], gains[i].Gx));
		}
	}
}